esp32_ulp/assemble.py

```diff
@@ -108,17 +108,21 @@ def parse_line(self, line):
         """
         if not line:
             return
-        has_label = line[0] not in '\t .'
+        has_label = ':' in line
```

Comment:
did you check if there is no ":" as part of the assembler syntax, except for labels?

we could also do a regex match like `^\S+:` or so.

Reply:
with named groups, this could even directly extract the label and the rest of the line and strip blanks.

Reply:
I was at one point thinking to rename the variable because it technically no longer checks if the line "has a label" but rather, if "it might have a label". In case where the colon means something else (e.g. when it's part of a string value), then the algorithm further down backs out of considering the text before the colon a label. (the strip with the long list of characters).

I didnt consider regex here, but I think it could maybe make the logic here easier to understand/read. I'll give a regex version of the same a go.

Reply:
I just tested a fair bit with a regex approach, because I started with the assumption, that it would be slower and use more memory than the handcrafted approach.

I compiled the regex only once, and then reused that compiled regex for every line.

Firstly about memory usage:
When running a single "parse_line" from the unit tests, the regex approach would use about 1.4k of memory more than the previous approach. However, when running a garbage collect right after, it would shrink to 64 bytes. Even running the parse_line many times over many other lines, the extra overhead stayed at 64 bytes (after garbage collect). I would consider this entirely negligible (i tried this on both the PC and the ESP32 and the overhead is 64 bytes in both cases).

Now for performance:
To test, I put a single "parse_line" from the unit tests into a loop, with 100k iterations (on the PC) and 10k iterations (on the ESP32) and measured the duration. On the PC, the regex approach take almost twice as long as the previous code (1527ms vs 726ms - averaged over 10 runs of 100k iterations each). This appeared to confirm my "it would be slower" suspicion, however that translates into only about 8us per line of code, which seems entirely negligible for the size of things that would ever be assembled.

To my surprise the performance was different on the ESP32. Here, with 10k iterations, the total time was very similar for both approaches, with the regex approach being even slightly faster. The duration for 10k iterations was around 43 seconds, and the regex version was between 1 and 2 seconds faster every time. I don't know why this would be the case, but I think it also supports the conclusion that the performance difference is not an issue, when switching to a regex approach.

So, regex approach it is. I will clean it up and push it some time later.

Reply:
Oh, that's a nice surprise.

When thinking about the regex approach again, I remembered the comment-removal code which I had to reinvent with a state machine because the regex implementation on micropython was too limited. Hopefully that isn't the case here also.